azure_costs_exporter/scrape.py

```diff
@@ -53,17 +53,27 @@ def convert_json_df(data):
 
     return df
 
-def extract_metrics_from_df(df, counter):
+def extract_metrics_from_df(df, metric_name, enrollment, registry):
     """
     Fill a counter object with the data from the data frame
     """
     groups = df.groupby(base_columns).sum()
-    for name, value in groups.iterrows():
-        meta = dict(zip(base_columns, name))
-        counter.record(value.ExtendedCost, **meta)
 
+    class AzureCollector(object):
+        def collect(self):
+            c = CounterMetricFamily(metric_name,
+                                    "Costs billed to Azure Enterprise Agreement {}".format(enrollment),
+                                    labels=base_columns)
 
-def query_metrics(enrollment, token, metric_name, month=None):
+            for name, value in groups.iterrows():
+                c.add_metric(name, value.ExtendedCost)
+
+            yield c
+
+    registry.register(AzureCollector())
```

Comment:
This looks wrong. Registration should be a one-time thing, but you would register the metrics anew every time `extract_metrics_from_df` is called.

It also seems very odd that your collector is a class that is nested within a function. I would expect anything that is related to collecting the metrics to be done in the `collect()` function, including gathering the data. You only do some postprocessing of data that may have been fetched a long time ago.

I do not know how this combines with the multiple registration thing, but I would not be too surprised if there were multiple instances (with multiple `#HELP` entries) of the same data with conflicting values (by the way, Prometheus would fail to collect such data).